File: CS/BarsMenuFromDataTable/Form1.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Windows.Forms;
using DevExpress.XtraBars;
using DevExpress.XtraBars.Styles;

namespace BarsMenuFromDataTable
{
	public partial class Form1 : Form
	{
		public Form1()
		{
			InitializeComponent();
		}

		private BarItem[] CreateChildrenLinks(BarItem parentItem, DataTable childrenTable)
		{
			List<BarItem> items = new List<BarItem>();
			foreach ( DataRow row in childrenTable.Rows )
			{
				BarItem barItem = null;

				int childrenCount = childrenTable.AsEnumerable().Count(r => (int)r["ParentId"] == (int)row["Id"]);
				if ( childrenCount > 0 )
				{
					barItem = new BarSubItem();
					DataTable table = childrenTable.AsEnumerable().Where(r => (int)r["ParentId"] == (int)row["Id"]).CopyToDataTable();
					((BarSubItem)barItem).AddItems(CreateChildrenLinks(barItem, table));
				}
				else
				{
					BarItemInfo itemInfo = (BarItemInfo)row["ItemKind"];
					barItem = System.Activator.CreateInstance(itemInfo.ItemType) as BarItem;
				}

				barItem.Name = ((string)row["Caption"]).ToLower();
				barItem.Caption = (string)row["Caption"];
				barItem.Manager = barManager1;

				items.Add(barItem);
			}

			return items.ToArray();
		}

		private void Form1_Load(object sender, EventArgs e)
		{
			DataTable table = CreateTable(30);
			gridControl1.DataSource = table;

			foreach ( DataRow row in table.Select("ParentId = -1", "Id asc") )
			{
				BarItem barItem = null;

				int childrenCount = table.AsEnumerable().Count(r => (int)r["ParentId"] == (int)row["Id"]);
				if ( childrenCount > 0 )
				{
					barItem = new BarSubItem();
					DataTable childrenTable = table.AsEnumerable().Where(r => (int)r["ParentId"] == (int)row["Id"]).CopyToDataTable();
					((BarSubItem)barItem).AddItems(CreateChildrenLinks(barItem, childrenTable));
				}
				else
				{
					BarItemInfo itemInfo = (BarItemInfo)row["ItemKind"];
					barItem = System.Activator.CreateInstance(itemInfo.ItemType) as BarItem;
				}

				barItem.Name = ((string)row["Caption"]).ToLower();
				barItem.Caption = (string)row["Caption"];
				barItem.Manager = barManager1;

				bar1.AddItem(barItem);
			}
		}

		private DataTable CreateTable(int RowCount)
		{
			BarManagerPaintStyle paintStyle = barManager1.GetController().PaintStyle;
			BarItemInfoCollection itemsCollection = new BarItemInfoCollection(paintStyle);

			for ( int i = 0; i < 5; i++ )
				if ( paintStyle.ItemInfoCollection[i].ItemType != typeof(BarSubItem) )
					itemsCollection.Add(paintStyle.ItemInfoCollection[i]);

			DataTable tbl = new DataTable();
			tbl.Columns.Add("Id", typeof(int));
			tbl.Columns.Add("ParentId", typeof(int));
			tbl.Columns.Add("Caption", typeof(string));
			tbl.Columns.Add("ItemKind", typeof(BarItemInfo));

			Random rnd = new Random();
			for ( int i = 0; i < RowCount; i++ )
			{
				int parentId = rnd.Next(RowCount * -1, RowCount);
				if ( parentId == i )
					parentId--;

				tbl.Rows.Add(new object[] { i + 1, parentId <= 0 ? -1 : parentId , String.Format("Caption {0}", i),
											itemsCollection[rnd.Next(0, itemsCollection.Count - 1)] });
			}
						
			return tbl;
		}
	}
}

File: VB/BarsMenuFromDataTable/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Linq
Imports System.Windows.Forms
Imports DevExpress.XtraBars
Imports DevExpress.XtraBars.Styles
Imports Microsoft.VisualBasic

Namespace BarsMenuFromDataTable
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private Function CreateChildrenLinks(ByVal parentItem As BarItem, ByVal childrenTable As DataTable) As BarItem()
			Dim items As New List(Of BarItem)()
			For Each row As DataRow In childrenTable.Rows
				Dim barItem As BarItem = Nothing

				Dim childrenCount As Integer = childrenTable.AsEnumerable().Count(Function(r) CInt(Fix(r("ParentId"))) = CInt(Fix(row("Id"))))
				If childrenCount > 0 Then
					barItem = New BarSubItem()
					Dim table As DataTable = childrenTable.AsEnumerable().Where(Function(r) CInt(Fix(r("ParentId"))) = CInt(Fix(row("Id")))).CopyToDataTable()
					CType(barItem, BarSubItem).AddItems(CreateChildrenLinks(barItem, table))
				Else
					Dim itemInfo As BarItemInfo = CType(row("ItemKind"), BarItemInfo)
					barItem = TryCast(System.Activator.CreateInstance(itemInfo.ItemType), BarItem)
				End If

				barItem.Name = (CStr(row("Caption"))).ToLower()
				barItem.Caption = CStr(row("Caption"))
				barItem.Manager = barManager1

				items.Add(barItem)
			Next row

			Return items.ToArray()
		End Function

		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			Dim table As DataTable = CreateTable(30)
			gridControl1.DataSource = table

			For Each row As DataRow In table.Select("ParentId = -1", "Id asc")
				Dim barItem As BarItem = Nothing

				Dim childrenCount As Integer = table.AsEnumerable().Count(Function(r) CInt(Fix(r("ParentId"))) = CInt(Fix(row("Id"))))
				If childrenCount > 0 Then
					barItem = New BarSubItem()
					Dim childrenTable As DataTable = table.AsEnumerable().Where(Function(r) CInt(Fix(r("ParentId"))) = CInt(Fix(row("Id")))).CopyToDataTable()
					CType(barItem, BarSubItem).AddItems(CreateChildrenLinks(barItem, childrenTable))
				Else
					Dim itemInfo As BarItemInfo = CType(row("ItemKind"), BarItemInfo)
					barItem = TryCast(System.Activator.CreateInstance(itemInfo.ItemType), BarItem)
				End If

				barItem.Name = (CStr(row("Caption"))).ToLower()
				barItem.Caption = CStr(row("Caption"))
				barItem.Manager = barManager1

				bar1.AddItem(barItem)
			Next row
		End Sub

		Private Function CreateTable(ByVal RowCount As Integer) As DataTable
			Dim paintStyle As BarManagerPaintStyle = barManager1.GetController().PaintStyle
			Dim itemsCollection As New BarItemInfoCollection(paintStyle)

			For i As Integer = 0 To 4
				If paintStyle.ItemInfoCollection(i).ItemType IsNot GetType(BarSubItem) Then
					itemsCollection.Add(paintStyle.ItemInfoCollection(i))
				End If
			Next i

			Dim tbl As New DataTable()
			tbl.Columns.Add("Id", GetType(Integer))
			tbl.Columns.Add("ParentId", GetType(Integer))
			tbl.Columns.Add("Caption", GetType(String))
			tbl.Columns.Add("ItemKind", GetType(BarItemInfo))

			Dim rnd As New Random()
			For i As Integer = 0 To RowCount - 1
				Dim parentId As Integer = rnd.Next(RowCount * -1, RowCount)
				If parentId = i Then
					parentId -= 1
				End If

				tbl.Rows.Add(New Object() { i + 1,If(parentId <= 0, -1, parentId), String.Format("Caption {0}", i), itemsCollection(rnd.Next(0, itemsCollection.Count - 1)) })
			Next i

			Return tbl
		End Function
	End Class
End Namespace
